File: solutions/javascript/dfs/battleshipsInABoard.js
const BATTLESHIP_CELL = 'X';
const directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
];

const getPositionString = (row, col) => `${row}, ${col}`;

const isInBounds = (grid, row, col) => {
    const numRows = grid.length;
    const numCols = grid[0].length;

    const rowInBounds = row >= 0 && row < numRows;
    const colInBounds = col >= 0 && col < numCols;

    return rowInBounds && colInBounds;
};

const markBattleshipAsVisited = (row, col, board, visited) => {
    // Base case
    if (!isInBounds(board, row, col)) return;

    const curPositionString = getPositionString(row, col);
    if (visited.has(curPositionString)) return;

    if (board[row][col] !== BATTLESHIP_CELL) return;

    // Process node
    visited.add(curPositionString);

    // Recurse on potential neighbors
    for (const direction of directions) {
        const [rowChange, colChange] = direction;

        const newRow = row + rowChange;
        const newCol = col + colChange;

        markBattleshipAsVisited(newRow, newCol, board, visited);
    }
};

const countBattleships = (board) => {
    const numRows = board.length;
    const numCols = board[0].length;

    let numBattleships = 0;
    const visited = new Set();

    for (let row = 0; row < numRows; row++) {
        for (let col = 0; col < numCols; col++) {
            if (board[row][col] !== BATTLESHIP_CELL) continue;

            const curPositionString = getPositionString(row, col);
            if (visited.has(curPositionString)) continue;

            numBattleships += 1;
            markBattleshipAsVisited(row, col, board, visited);
        }
    }

    return numBattleships;
};

File: solutions/python/dfs/battleships_in_a_board.py
BATTLESHIP_CELL = 'X'
directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
]

def is_in_bounds(grid, row, col):
    num_rows = len(grid)
    num_cols = len(grid[0])

    row_in_bounds = row >= 0 and row < num_rows
    col_in_bounds = col >= 0 and col < num_cols

    return row_in_bounds and col_in_bounds

class Solution:
    def mark_battleship_as_visited(self, row, col, board, visited):
        # Base case
        if not is_in_bounds(board, row, col):
            return

        if (row, col) in visited:
            return

        if board[row][col] != BATTLESHIP_CELL:
            return

        # Process node
        visited.add((row, col))

        # Recurse on potential neighbors
        for direction in directions:
            row_change, col_change = direction

            new_row = row + row_change
            new_col = col + col_change

            self.mark_battleship_as_visited(new_row, new_col, board, visited)

    def countBattleships(self, board: List[List[str]]) -> int:
        num_rows = len(board)
        num_cols = len(board[0])

        num_battleships = 0
        visited = set()

        for row in range(num_rows):
            for col in range(num_cols):
                if board[row][col] != BATTLESHIP_CELL:
                    continue

                if (row, col) in visited:
                    continue

                num_battleships += 1
                self.mark_battleship_as_visited(row, col, board, visited)

        return num_battleships
